json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/JsonSchemaUtils.java

```diff
@@ -281,79 +282,90 @@ public static byte[] toJson(Object value) throws IOException {
           fieldNameBuilder.append(".");
-          fieldNameBuilder.append(identifiersList.poll());
+          word = identifiersList.poll();
+          if (word == null) {
+            return null;
+          }
+          fieldNameBuilder.append(word);
         }
         JsonNode fieldNode = fields.get(fieldNameBuilder.toString());
         if (fieldNode == null || identifiersList.isEmpty()) {
           return fieldNode;
         } else {
-          return findNodeHelper(fieldNode, identifiersList);
+          return findNodeHelper(fieldNode, entityType, identifiersList);
         }
       case "array":
-        return findNodeHelper(node.get("items"), identifiersList);
+        return findNodeHelper(node.get("items"), entityType, identifiersList);
       case "anyof":
         if (node.has("anyOf")) {
-          return findSubSchemasFromIndex(node, "anyOf", identifiersList);
+          return findSubSchemasFromIndex(node, entityType, "anyOf", identifiersList);
         }
 
         // handle "type": ["null", "object"] etc
         identifiersList.poll();
-        return findNodeHelper(node, identifiersList);
+        return findNodeHelper(node, entityType, identifiersList);
       case "allof":
         if (node.has("allOf")) {
-          return findSubSchemasFromIndex(node, "allOf", identifiersList);
+          return findSubSchemasFromIndex(node, entityType, "allOf", identifiersList);
         }
 
         // handle if-then-else
         identifiersList.poll();
-        if (identifiersList.peek().equals("conditional")) {
+        if ("conditional".equals(identifiersList.peek())) {
           identifiersList.poll();
-          return findNodeHelper(node.get(identifiersList.poll()), identifiersList);
+          return findNodeHelper(node.get(identifiersList.poll()), entityType, identifiersList);
         } else {
-          return findNodeHelper(node, identifiersList);
+          return findNodeHelper(node, entityType, identifiersList);
         }
       case "oneof":
-        return findSubSchemasFromIndex(node, "oneOf", identifiersList);
+        return findSubSchemasFromIndex(node, entityType, "oneOf", identifiersList);
       case "not":
-        return findNodeHelper(node.get(type), identifiersList);
+        return findNodeHelper(node.get(type), entityType, identifiersList);
       default:
         return null;
     }
   }
 
-  private static JsonNode findSubSchemasFromIndex(JsonNode node, String type,
+  private static JsonNode findSubSchemasFromIndex(JsonNode node,
+                                                  SchemaEntity.EntityType entityType,
+                                                  String compositionType,
                                                   LinkedList<String> identifiersList) {
     String index = identifiersList.poll();
-    List<JsonNode> jsonNodeList = Lists.newArrayList(node.get(type).elements());
+    List<JsonNode> jsonNodeList = Lists.newArrayList(node.get(compositionType).elements());
```

Comment:
nit:  add `if (index == null) { return null; }` after `String index = identifiersList.poll()` to avoid creating the lists unnecessarily